### PR TITLE
Make type hashing `:total`

### PR DESCRIPTION
**Pull request by Seelengrab**
Since types can't really be deleted and the hash of a type is already being cached inside of the type object itself, it should be legal to mark the `ccall` retrieving that hash as `:total`, permitting hashing of types to be concretely evaluated.

I'm unsure if the test I've added is good as written, or whether this should be done differently.

### CODE REVIEW

**Comment by Tokazama** [comment]
I thought the hash of data types [could change between runs](https://docs.julialang.org/en/v1/manual/modules/#Module-initialization-and-precompilation)

**Comment by vtjnash** [comment]
That was fixed on master recently, so we should remove that documentation

File: test/hashing.jl [hunk 1]
```diff
@@ -302,3 +302,5 @@ struct AUnionParam{T<:Union{Nothing,Float32,Float64}} end
     # test hashing of rational with odd denominator
     @test hash(5//3) == hash(big(5)//3)
 end
+
+@test Base.infer_effects(hash, Tuple{Type{Int}, UInt}) == Core.Compiler.EFFECTS_TOTAL
```
Comment (aviatesk) [inline test/hashing.jl:306]:
```suggestion
@test Core.Compiler.is_foldable_nothrow(Base.infer_effects(hash, Tuple{Type{Int}, UInt}))
```
is the way to test this.

Comment (Seelengrab) [reply test/hashing.jl:306]: Thank you!

**Comment by Seelengrab** [comment]
I think "run to run" in that documentation refers to different invocations of julia/"sessions of the REPL", not different queries of `hash`, so that should already be consistent.

File: test/hashing.jl [hunk 2]
```diff
@@ -302,3 +302,5 @@ struct AUnionParam{T<:Union{Nothing,Float32,Float64}} end
     # test hashing of rational with odd denominator
     @test hash(5//3) == hash(big(5)//3)
 end
+
+@test Core.Compiler.is_foldable_nothrow(hash, Tuple{Type{Int}, UInt})
```
Comment (aviatesk) [inline test/hashing.jl:306]:
```suggestion
@test Core.Compiler.is_foldable_nothrow(Base.infer_effects(hash, Tuple{Type{Int}, UInt}))
```

File: base/hashing.jl [hunk 3]
```diff
@@ -29,7 +29,10 @@ See also: [`objectid`](@ref), [`Dict`](@ref), [`Set`](@ref).
 """
 hash(x::Any) = hash(x, zero(UInt))
 hash(w::WeakRef, h::UInt) = hash(w.value, h)
-hash(T::Type, h::UInt) = hash_uint(3h - ccall(:jl_type_hash, UInt, (Any,), T))
+
+# Types can't be deleted, so marking as total allows the compiler to look up the hash
+@assume_effects :total _typehash(T::Type) = ccall(:jl_type_hash, UInt, (Any,), T)
+hash(T::Type, h::UInt) = hash_uint(3h - _typehash(T))
```
[the remark this answers is not in the document]
Comment (Seelengrab) [reply base/hashing.jl:35]: Ah nice! I saw the PR, but didn't know it was merged already.

**Comment by topolarity** [comment]
@vtjnash Do you know if the hashing changes you mention on master were backported to 1.10?

Would be nice to backport this as well, if so

**Comment by oscardssmith** [comment]
This wasn't backported (and it would be a somewhat risky backport imo https://github.com/JuliaLang/julia/pull/52119)

**Pull request merged**